File: ValidadorPolinomios/src/Polinomio/Expressao.java
package Polinomio;

import java.util.ArrayList;
import java.util.HashMap;

public class Expressao extends Base {
	private ArrayList<Termo> termos;
	private Termo ultimoTermo;
	private Base origem;

	public Expressao() {
		termos = new ArrayList<>();
	}

	public void addTermo(Termo termo) {
		ultimoTermo = termo;
		termo.setOrigem(this);
		termos.add(termo);
	}

	public ArrayList<Termo> getTermos() {
		return termos;
	}

	public Termo getUltimoTermo() {
		return ultimoTermo;
	}

	public double calcular(HashMap<Character, Integer> variaveis) {
		double result = 0;
		for (Termo termo : termos) {
			result += termo.calcular(variaveis);
		}
		return result;
	}

	public Base getOrigem() {
		return origem;
	}

	public void setOrigem(Base origem) {
		this.origem = origem;
	}

	@Override
	public String toString(boolean traduzido, HashMap<Character, Integer> variaveis) {
		String result = "";
		for (int i = 0; i < termos.size(); i++) {
			if (i > 0)
				result += "+";
			result += termos.get(i).toString(traduzido, variaveis);
		}
		return result;
	}

	public Expressao simplificar() {
		for (int i = 0; i < termos.size(); i++) {
			termos.get(i).simplificar();
		}
		return this;
	}

	public void substituir(Termo termoSubstituido, ArrayList<Termo> termosSubstituidores) {
		int indexTermoSubstituido = this.termos.indexOf(termoSubstituido);
		this.termos.remove(indexTermoSubstituido);

		for (Termo termoSubstituidor : termosSubstituidores) {
			Sinal novoSinal = Sinal
					.valueOf(termoSubstituidor.getSinal().getValue() * termoSubstituido.getSinal().getValue());
			termoSubstituidor.setSinal(novoSinal);
			termoSubstituidor.setOrigem(this);
			this.termos.add(indexTermoSubstituido++, termoSubstituidor);
		}
	}

	@Override
	public Base ordenar() {

		for (Termo termo : termos) {
			termo.ordenar();
		}
		
		this.ordernar(termos);
		
		/*Termo termoAux = null;
		boolean teveTroca = false;
		int max = termos.size() - 1;
		for (int i = 0; i < termos.size(); i++) {
			teveTroca = false;
			for (int j = 0; j < max; j++) {
				if (termos.get(j).getPeso() > termos.get(j + 1).getPeso()) {
					termoAux = termos.remove(j);
					termos.add(termos.get(j + 1));
					termos.add(j + 1, termoAux);
					teveTroca = true;
				}
			}
			if (!teveTroca)
				break;
			max--;
		}*/

		return this;
	}

	@Override
	public int getPeso() {
		int total = 0;
		for (Termo termo : termos) {
			total += termo.getPeso();
		}
		return total;
	}
}

File: ValidadorPolinomios_CS/ValidadorPolinomios/ValidadorPolinomios/Polinomio/Expressao.cs
using System.Collections.Generic;

namespace ValidadorPolinomios.Polinomio
{
    public class Expressao : Base
    {
        private List<Termo> termos;
        private Termo ultimoTermo;
        private Base origem;

        public Expressao()
        {
            termos = new List<Termo>();
        }

        public virtual void addTermo(Termo termo)
        {
            ultimoTermo = termo;
            termo.Origem = this;
            termos.Add(termo);
        }

        public virtual List<Termo> Termos
        {
            get
            {
                return termos;
            }
        }

        public virtual Termo UltimoTermo
        {
            get
            {
                return ultimoTermo;
            }
        }

        public override double calcular(Dictionary<char, int> variaveis)
        {
            double result = 0;
            foreach (Termo termo in termos)
            {
                result += termo.calcular(variaveis);
            }
            return result;
        }

        public override string ToString(bool traduzido, Dictionary<char, int> variaveis)
        {
            string result = "";
            for (int i = 0; i < termos.Count; i++)
            {
                if (i > 0)
                {
                    result += "+";
                }
                result += termos[i].ToString(traduzido, variaveis);
            }
            return result;
        }

        public override Base simplificar()
        {
            for (int i = 0; i < termos.Count; i++)
            {
                termos[i].simplificar();
            }
            return this;
        }

        public virtual void substituir(Termo termoSubstituido, List<Termo> termosSubstituidores)
        {
            int indexTermoSubstituido = this.termos.IndexOf(termoSubstituido);
            this.termos.RemoveAt(indexTermoSubstituido);

            foreach (Termo termoSubstituidor in termosSubstituidores)
            {
                Sinal novoSinal = Sinal.valueOf(termoSubstituidor.Sinal.Value * termoSubstituido.Sinal.Value);
                termoSubstituidor.Sinal = novoSinal;
                termoSubstituidor.Origem = this;
                this.termos.Insert(indexTermoSubstituido++, termoSubstituidor);
            }
        }

        public override Base ordenar()
        {
            foreach (Termo termo in termos)
                termo.ordenar();

            this.ordernar(termos);
            return this;
        }

        public override int Peso
        {
            get
            {
                int total = 0;
                foreach (Termo termo in termos)
                {
                    total += termo.Peso;
                }
                return total;
            }
        }
    }

}